tracer/src/Datadog.Trace/Activity/ActivityListener.cs

```diff
@@ -0,0 +1,289 @@
+            {
+                return null;
+            }
+
+            return activity.DuckAs<IActivity6>() ??
+                   activity.DuckAs<IActivity5>() ??
+                   activity.DuckAs<IW3CActivity>() ??
+                   activity.DuckAs<IActivity>();
+        }
+
+        public static void Initialize() => Initialize(CancellationToken.None);
+
+        public static void Initialize(CancellationToken cancellationToken)
+        {
+            if (Interlocked.CompareExchange(ref _initialized, 1, 0) == 1)
+            {
+                return;
+            }
+
+            // Try to resolve System.Diagnostics.DiagnosticListener, System.Diagnostics.DiagnosticSource
+            var diagnosticListenerType = Type.GetType("System.Diagnostics.DiagnosticListener, System.Diagnostics.DiagnosticSource", throwOnError: false);
+            if (diagnosticListenerType is null)
+            {
+                // Because we cannot resolve the DiagnosticListener type we allow a later initialization.
+                // For this case we are going to do some retries with a back-off.
+                if (Interlocked.Decrement(ref _initializationRetries) > 0)
+                {
+                    Task.Delay(InitializationBackoffPerRetry, cancellationToken).ContinueWith(
+                        _ =>
+                    {
+                        if (cancellationToken.IsCancellationRequested)
+                        {
+                            return;
+                        }
+
+                        Interlocked.Exchange(ref _initialized, 0);
+                        Initialize(cancellationToken);
+                    },
+                        cancellationToken,
+                        TaskContinuationOptions.None,
+                        TaskScheduler.Default);
+                }
+
+                return;
+            }
+
+            // Initialize
+            var diagnosticSourceAssemblyName = diagnosticListenerType.Assembly.GetName();
+            Log.Information("DiagnosticSource: {diagnosticSourceAssemblyNameFullName}", diagnosticSourceAssemblyName.FullName);
+
+            var activityType = Type.GetType("System.Diagnostics.Activity, System.Diagnostics.DiagnosticSource", throwOnError: false);
+            var version = diagnosticSourceAssemblyName.Version;
+
+            // Check version: First version where the Activity objects has traceId and spanId
+            if (version >= new Version(4, 0, 4))
+            {
+                if (activityType is null)
+                {
+                    Log.Error("The Activity type cannot be found.");
+                    return;
+                }
+
+                CreateCurrentActivityDelegates(activityType);
+                ChangeActivityDefaultFormat(activityType);
+
+                if (version.Major >= 5)
+                {
+                    // if Version >= 5 loaded (Uses ActivityListener implementation)
+                    CreateActivityListenerInstance(activityType);
```

Comment:
Perhaps a silly question, but if the `System.Diagnostics.DiagnosticSource.dll` is new enough to have ActivityListener's but a library is emitting `DiagnosticSource` events, will we still catch them?

Reply:
Nope, we are only catching activities, so I'm using the way each version handles notifications. We will miss any other custom object reported by DiagnosticSource.